src/mo_idl/mo_to_idl.ml

```diff
@@ -44,74 +45,88 @@ module MakeState() = struct
              stamp := Stamp.add c (k, n + 1) !stamp;
              type_map := TypeMap.add id (k, n + 1) !type_map;
              (k, n + 1))
-        | Some (k, n) -> (k, n)
+        | Some kn -> kn
       in
       begin
         match (k, n) with
+        | _ when k < 0 || n < 0 -> assert false
         | (0, 0) -> name
-        | (0, n) when n > 0 -> Printf.sprintf "%s_%d" name n
-        | (k, 0) when k > 0 -> Printf.sprintf "%s__%d" name k
-        | (k, n) when k > 0 && n > 0 -> Printf.sprintf "%s__%d_%d" name k n
-        | _ -> assert false
+        | (0, n) -> Printf.sprintf "%s_%d" name n
+        | (k, 0) -> Printf.sprintf "%s__%d" name k
+        | (k, n) -> Printf.sprintf "%s__%d_%d" name k n
       end
     | _ -> assert false
 
-  let prim = function
-    | Null -> I.PrimT I.Null
-    | Bool -> I.PrimT I.Bool
-    | Nat -> I.PrimT I.Nat
-    | Nat8 -> I.PrimT I.Nat8
-    | Nat16 -> I.PrimT I.Nat16
-    | Nat32 -> I.PrimT I.Nat32
-    | Nat64 -> I.PrimT I.Nat64
-    | Int -> I.PrimT I.Int
-    | Int8 -> I.PrimT I.Int8
-    | Int16 -> I.PrimT I.Int16
-    | Int32 -> I.PrimT I.Int32
-    | Int64 -> I.PrimT I.Int64
-    | Float -> I.PrimT I.Float64
-    | Char -> I.PrimT I.Nat32
-    | Text -> I.PrimT I.Text
-    | Blob -> I.BlobT
-    | Principal -> I.PrincipalT
+  let prim = let open I in
+    function
+    | Type.Null -> PrimT Null
+    | Bool -> PrimT Bool
+    | Nat -> PrimT Nat
+    | Nat8 -> PrimT Nat8
+    | Nat16 -> PrimT Nat16
+    | Nat32 -> PrimT Nat32
+    | Nat64 -> PrimT Nat64
+    | Int -> PrimT Int
+    | Int8 -> PrimT Int8
+    | Int16 -> PrimT Int16
+    | Int32 -> PrimT Int32
+    | Int64 -> PrimT Int64
+    | Float -> PrimT Float64
+    | Char -> PrimT Nat32
+    | Text -> PrimT Text
+    | Blob -> BlobT
+    | Principal -> PrincipalT
     | Region
     | Error -> assert false
 
-  let rec typ t =
-    (match t with
-    | Any -> I.PrimT I.Reserved
-    | Non -> I.PrimT I.Empty
+  let rec typ t0 =
+    (match t0 with
+    | Any -> I.(PrimT Reserved)
+    | Non -> I.(PrimT Empty)
     | Prim p -> prim p
     | Var (s, i) -> assert false
     | Con (c, ts) ->
        (match Cons.kind c with
         | Def (_, t) ->
            (match (open_ ts t) with
             | Prim p -> prim p
-            | Any -> I.PrimT I.Reserved
-            | Non -> I.PrimT I.Empty
+            | Any -> I.(PrimT Reserved)
+            | Non -> I.(PrimT Empty)
+            | Con (c1, ts1) when ts = [] && Cons.eq c c1 -> failwith "CON EQ"
+            | Con (c1, ts1) when false && List.length ts = 0 && List.length ts1 = 1 && Cons.name c = Cons.name c1 -> failwith (Cons.name c ^ " = Innerer.Credit<Nat>")
+            | Con (c1, ts1) when false && List.length ts = 1 && compare ts ts1 = 0 && Cons.name c = Cons.name c1 -> failwith "CON"
+
+
+            | Con (c1, ts1) when compare ts ts1 = 0 && Cons.name c = Cons.name c1 ->
```

Comment:
```suggestion
            | Con (c1, ts1) when ts = ts1 && Cons.name c = Cons.name c1 ->
```
would this work?

Reply:
I don't think so. Perhaps a better way would be to check the def ( not the expansion) is 
<T0, ... TN> d<T0,..., Tn>
ie
(vs, Con(c1, [Var n,.... Var 0])  where len vs = n, or maybe
(vs, Con(c1, [Var 0,.... Var n])  where len vs = n
depending on how we number our deBruin variables.

Then you would know that c1 is an eta-expansion of c2, probably arising from
type c<T1,....Tn> = c1<T1,..., Tn>

If you checked the names (not stamps) of the constructors are eqaul then you can probably just shorted
type Result<T,E> = Result.Result<T,E>
but not
type ErrorMonad<T, E> = Resut.Result<T,E>

Stupid question, but does Yan's code monomorphize 0-ary constructor applications. That might be silly...

Reply:
Abandoned this approach, see PR's description.

Reply:
Btw. I think also 0-ary `Con's need to be "monomorphised", as there already may be a homonymous perviously discovered n-ary `Con` in the `env`.